src/field_theory/splitting_field.lean

```diff
@@ -92,63 +93,409 @@ theorem nat_degree_remove_factor' {f : K[X]} {n : ℕ} (hfn : f.nat_degree = n+1
+
+protected lemma nsmul_zero (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (x : splitting_field_aux n f), (0 : ℕ) • x = splitting_field_aux.zero n :=
+nat.rec_on n (λ K fK f, by exactI zero_nsmul) (λ n ih K fK f, by exactI ih)
+
+protected lemma nsmul_succ (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (k : ℕ) (x : splitting_field_aux n f),
+  (k + 1) • x = splitting_field_aux.add n x (k • x) :=
+nat.rec_on n (λ K fK f n x, by exactI succ_nsmul x n) (λ n ih K fK f, by exactI ih)
+
+protected lemma zsmul_zero (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (x : splitting_field_aux n f), (0 : ℤ) • x = splitting_field_aux.zero n :=
+nat.rec_on n (λ K fK f, by exactI zero_zsmul) (λ n ih K fK f, by exactI ih)
+
+protected lemma zsmul_succ (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (k : ℕ) (x : splitting_field_aux n f),
+  (↑(k + 1) : ℤ) • x = splitting_field_aux.add n x ((k : ℤ) • x) :=
+nat.rec_on n (λ K fK f k x, by exactI @subtraction_comm_monoid.zsmul_succ' K _ k x)
+  (λ n ih K fK f, by exactI ih)
+
+protected lemma zsmul_neg (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (k : ℕ) (x : splitting_field_aux n f),
+  -[1+ k] • x = splitting_field_aux.neg n ((k + 1 : ℤ) • x) :=
+nat.rec_on n (λ K fK f k x, by exactI @subtraction_comm_monoid.zsmul_neg' K _ k x)
+  (λ n ih K fK f, by exactI ih)
+
+/-- `K` adjoined `n` roots of `f : K[X]` is an additive, commutative group.
+
+We need this instance to define the `distrib_mul_action` instance below; a `distrib_mul_action`
+is required to lift scalar multiplications on `K` to a splitting field of `f : K[X]`.
+-/
+instance add_comm_group (n : ℕ) {K : Type u} [field K] {f : K[X]} :
+  add_comm_group (splitting_field_aux n f) :=
+begin
+  refine_struct
+    { add := splitting_field_aux.add n,
+      zero := splitting_field_aux.zero n,
+      neg := splitting_field_aux.neg n,
+      sub := splitting_field_aux.sub n,
+      nsmul := (•),
+      nsmul_zero' := splitting_field_aux.nsmul_zero n,
+      nsmul_succ' := splitting_field_aux.nsmul_succ n,
+      zsmul := (•),
+      zsmul_zero' := splitting_field_aux.zsmul_zero n,
+      zsmul_succ' := splitting_field_aux.zsmul_succ n,
+      zsmul_neg' := splitting_field_aux.zsmul_neg n },
+  all_goals { unfreezingI { induction n with n ih generalizing K } },
+  -- The `succ` cases follow by induction, handle them first.
+  iterate 6 { rotate, exact ih },
+  -- The `zero` cases follow from the structure of the field `K`.
+  all_goals { intros,
+    dsimp only [splitting_field_aux, splitting_field_aux.add, splitting_field_aux.zero,
+      splitting_field_aux.neg, splitting_field_aux.sub] },
+  -- `simp` can handle a couple of them
+  any_goals { simp },
+  -- and `ring` can handle the rest.
+  all_goals { ring },
+end
+
+instance (n : ℕ) {K : Type u} [field K] {f : K[X]} : add_monoid (splitting_field_aux n f) :=
+by apply_instance
```

Comment:
during development I needed it to have a proof not time out (I think it was actually the `distrib_mul_action` one below) but it may just have been one of those things where Lean doesn't evaluate things correctly when all the info about the `def` isn't given - it's now fine! So removed the instance.